ungdev/UA-api
# Discord team locking messages (#164)

* refactor: removed routes for team locking
refactor: when a cart is paid, refunded or forced paid, teams can be locked or unlocked

* fix: bugs

fix: we didn't check that the tournament wasn't full before locking the team

fix: we didn't check that the team was full before locking team

fix: discord was not setup

* feat: added a queue for teams, so that they can be locked even if there is no place left
fix: there was a problem when fetching player tickets in updateCart() operation

* test: added tests

* added tests for route POST /etupay/callback
* removed tests for route POST /teams/current/lock

* test: added tests for route POST /admin/users/:userId/force-pay

* added the tests
* fixed a bug where a team could be not full but force pay would still lock it (or put it on the waiting list)

* fix: small corrections in the tests

* fixed some typos in the name of the tests of the file etupay/callback.test.ts
* added some await at the end of a test to make sure instructions have finished running

* test: added tests for route POST /admin/carts/:cartId/refund

* added tests
* fixed a bug in file operations/carts.ts where the check of the name of the item was done wrong

* style: linted

* feat: when a players arrives/leaves a team, it may now lock/unlock the team

* test: added tests for route DELETE /teams/current/users/current

* added tests
* changed signature of the kickUser method : it now takes the user and not only its id
* fixed a bug where any type of members made the team be unlocked when they were kicked / they left
* route DELETE /teams/current/users/current no longer needs the team to not be locked

* test: added tests for route DELETE /teams/current/users/:userId

* added tests
* route DELETE /teams/current/users/:userId no longer needs the team to not be locked

* test: tests for route POST /teams/current/join-requests/:userId

* added tests
* the team can now be locked to make the request (it will still return an error tho, but not due to the fact that it is locked, but instead that it is full)
* fixed bugs with the remake of functions kickUser and joinTeam (in the team operations file)

* style: linted

* test: route DELETE /teams/current

* added tests
* team can now be locked to delete it
* a team that is locked and deleted now lets its place to a team in the waiting list

* fix: teams never needs to not be locked anymore

* removed the middleware isTeamNotLocked, and its usages
* removed the tests that tested that middleware

* test: added some tests

* added tests to verify a team in the queue would be moved to the locked teams if necessary with routes :
	- DELETE /teams/current/users/:userId
	- DELETE /teams/current/users/current
	- POST /admin/carts/:cartId/refund

* style: lint

* linted code
* for operation imports : replaced single function imports by gobal imports

* feat: updated POST /admin/users/:userId/replace

* team can now be locked OR in the queue
* added test for that

* fix(tests): fix discord oauth to add support for nodejs 18+ crypto module

* fix: /tournaments route test
* a test was not passing because there was neither casters nor cashprizes on the tournaments

* fix: team lock when creating a solo team (tournament with one player team), and the player has paid

* fix: refreshed pnpm-lock.yaml

* fix: the lockfile should be useable now

* fix: tests (csgo was still used instead of cs2)

* fix: added more time in the timeout for tests POST /teams/current/join-requests/:userId

* feat: add discord messages (webhooks) when a team is locked/unlocked

* fix: tests + readded pnpm-lock.yaml

* fix: lint

* feat: count tournament places in discord locked messages

* fix: removed unnecessary this.timeout(30000) at the beginning of some tests

* fix: missing await + no need for player list when unlocking

* fix: removed unnecessary eslint disabled in acceptRequest.test.ts
